src/features/envVisualizer/EnvVisualizerLayout.tsx

```diff
@@ -30,25 +32,70 @@ export class Layout {
   static height: number;
   /** the width of the stage */
   static width: number;
+  /** the visible height of the stage */
+  static visibleHeight: number = window.innerHeight;
+  /** the visible width of the stage */
+  static visibleWidth: number = window.innerWidth;
+  /** total width of stage */
+  static stageHeight: number = window.innerHeight;
+  /** the visible width of the stage */
+  static stageWidth: number = window.innerWidth;
   /** the unique key assigned to each node */
   static key: number = 0;
 
   /** the environment tree */
   static environmentTree: EnvTree;
   /** the global environment */
   static globalEnvNode: EnvTreeNode;
-  /** array of levels, which themselves are arrays of frames */
-  static levels: Level[];
+  /** grid of frames */
+  static grid: Grid;
   /** memoized values */
   static values = new Map<Data, Value>();
+  static oldValues = new Map<Data, Value>();
   /** memoized layout */
   static prevLayout: React.ReactNode;
+  static currentDark: React.ReactNode;
+  static currentLight: React.ReactNode;
+  static stageRef: RefObject<any> = React.createRef();
+  // buffer for faster rendering of diagram when scrolling
+  static invisiblePaddingVertical: number = 300;
+  static invisiblePaddingHorizontal: number = 300;
+  static scrollContainerRef: RefObject<any> = React.createRef();
+
+  static updateDimensions(width: number, height: number) {
+    Layout.visibleWidth = width;
+    Layout.visibleHeight = height;
+    if (
+      Layout.stageRef.current !== null &&
+      (window.innerWidth > Layout.stageWidth || window.innerHeight > Layout.stageHeight)
+    ) {
+      Layout.currentLight = undefined;
+      Layout.currentDark = undefined;
+      Layout.stageWidth = Math.min(Layout.width, window.innerWidth);
+      Layout.stageHeight = Math.min(Layout.height, window.innerHeight);
+      Layout.stageRef.current.width(Layout.stageWidth);
+      Layout.stageRef.current.height(Layout.stageHeight);
+      EnvVisualizer.redraw();
+    }
+    Layout.invisiblePaddingVertical = (Layout.stageHeight - Layout.visibleHeight) / 2;
+    Layout.invisiblePaddingHorizontal = (Layout.stageWidth - Layout.visibleWidth) / 2;
+
+    const container: HTMLElement | null = this.scrollContainerRef.current as HTMLDivElement;
```

Comment:
is this necessary? noticed that there is already some inline styling for this container below, which takes on the values dynamically.
(i haven't had the chance to run the code yet 😅)

Reply:
Currently the dimensions of the div created in `static draw(): React.ReactNode` isn't set dynamically so this updateDimensions method is required to resize the diagram and the scroll container